### PR TITLE
PCHR-2191: Cancel available for past dates for staff

### PR DESCRIPTION
**Pull request by dhavalsharma**
## Overview
Cancel option was always available to staff.

## Before
Cancel option was always available to staff.

## After
Cancel option is available to staff based on absence type setting for `allow_request_cancelation`.

manager-actions

## Technical Details
Absence type `allow_request_cancelation` is checked along with `from_date` of leave request to decide if staff can cancel the request.

## Comments
NA

---

- [X] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 1]
```diff
@@ -87,6 +88,26 @@ define([
...
+     * Checks if user is allowed to cancel leave request. It is based on following constants
+     * const REQUEST_CANCELATION_NO = 1;
+     * const REQUEST_CANCELATION_ALWAYS = 2;
+     * const REQUEST_CANCELATION_IN_ADVANCE_OF_START_DATE = 3;
+     *
+     * @param  {Object} request
+     * @return {Boolean}
+     */
+    vm.canCancel = function (request) {
+      if (vm.absenceTypes[request.type_id].allow_request_cancelation == 3) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Can be simplified
``` javascript
var allow_request_cancelation = vm.absenceTypes[request.type_id].allow_request_cancelation;
 if (allow_request_cancelation == 3) {
   return moment().isBefore(request.from_date);
 } 
 
 return allow_request_cancelation == 2;

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 2]
```diff
@@ -868,6 +871,64 @@
         }
       });
 
+      describe('canCancel', function() {
+        var leaveRequest1;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: why 1? it should just `leaveRequest`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 3]
```diff
@@ -868,6 +871,64 @@
...
+
+        describe('when absence type does allow cancellation in advance of start date', function() {
+          beforeEach(function() {
+            leaveRequest1.type_id = absenceTypeData.findByKeyValue('allow_request_cancelation','3').id;
+          });
+
+          describe('when from date is less than today', function() {
+            beforeEach(function() {
+              var baseDate = moment(leaveRequest1.from_date);
+              var baseMorethanFromDate = baseDate.add(10, 'days').toDate();
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: Better name `dateAfterFromDate `

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 4]
```diff
@@ -868,6 +871,64 @@
...
+
+            it('does not allow user to cancel request', function() {
+              expect(controller.canCancel(leaveRequest1)).toBe(false);
+            });
+          });
+
+          describe('when from date is more than today', function() {
+            beforeEach(function() {
+              var baseDate = moment(leaveRequest1.from_date);
+              var baseMorethanFromDate = baseDate.subtract(10, 'days').toDate();
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: Its not more, it should be less. Better name dateBeforeFromDate

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 5]
```diff
@@ -1,7 +1,8 @@
 define([
   'leave-absences/my-leave/modules/components',
-  'common/lodash'
-], function (components, _) {
+  'common/lodash',
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Common dependencies should be first.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 6]
```diff
@@ -87,6 +88,26 @@ define([
     };
 
     /**
+     * Checks if user is allowed to cancel leave request. It is based on following constants
+     * const REQUEST_CANCELATION_NO = 1;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Where these constants are coming from?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js (hunk 5 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Did you use lodash? Remove if not.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js (hunk 6 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: backend

**Pull request merged**